src/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs

```diff
@@ -0,0 +1,150 @@
+        {
+            MethodInfo createRange = ImmutableDictionaryCreateRangeMethod(constructingType, elementType);
+
+            if (createRange == null)
+            {
+                return null;
+            }
+
+            return createRange.CreateDelegate(
+                typeof(DefaultImmutableConverter.ImmutableDictCreateRangeDelegate<,>).MakeGenericType(typeof(string), elementType), null);
+        }
+
+        public override Func<object, TProperty> CreatePropertyGetter<TClass, TProperty>(PropertyInfo propertyInfo) =>
+            (Func<object, TProperty>)CreatePropertyGetter(propertyInfo, typeof(TClass));
+
+        private static Delegate CreatePropertyGetter(PropertyInfo propertyInfo, Type classType)
+        {
+            MethodInfo realMethod = propertyInfo.GetGetMethod();
+            Type objectType = typeof(object);
+
+            var dynamicMethod = new DynamicMethod(
+                realMethod.Name,
+                propertyInfo.PropertyType,
+                new[] { objectType },
+                typeof(ReflectionEmitMemberAccessor).Module,
+                skipVisibility: true);
+
+            ILGenerator generator = dynamicMethod.GetILGenerator();
+
+            generator.Emit(OpCodes.Ldarg_0);
+
+            if (classType.IsValueType)
+            {
+                generator.Emit(OpCodes.Unbox_Any, classType);
+                generator.Emit(OpCodes.Call, realMethod);
+            }
+            else
+            {
+                generator.Emit(OpCodes.Castclass, classType);
+                generator.Emit(OpCodes.Callvirt, realMethod);
+            }
+
+            generator.Emit(OpCodes.Ret);
+
+            return dynamicMethod.CreateDelegate(typeof(Func<,>).MakeGenericType(objectType, propertyInfo.PropertyType));
+        }
+
+        public override Action<object, TProperty> CreatePropertySetter<TClass, TProperty>(PropertyInfo propertyInfo) =>
+            (Action<object, TProperty>)CreatePropertySetter(propertyInfo, typeof(TClass));
+
+        private static Delegate CreatePropertySetter(PropertyInfo propertyInfo, Type classType)
+        {
+            MethodInfo realMethod = propertyInfo.GetSetMethod();
+            Type objectType = typeof(object);
+
+            var dynamicMethod = new DynamicMethod(
+                realMethod.Name,
+                typeof(void),
+                new[] { objectType, propertyInfo.PropertyType },
+                typeof(ReflectionEmitMemberAccessor).Module,
+                skipVisibility: true);
+
+            ILGenerator generator = dynamicMethod.GetILGenerator();
+
+            generator.Emit(OpCodes.Ldarg_0);
+
+            if (classType.IsValueType)
+            {
+                generator.Emit(OpCodes.Unbox_Any, classType);
+                generator.Emit(OpCodes.Ldarg_1);
+                generator.Emit(OpCodes.Call, realMethod);
+            }
+            else
+            {
+                generator.Emit(OpCodes.Castclass, classType);
```

Comment:
Originally was using emit for getters\setters, then [switched ](https://github.com/dotnet/corefx/pull/35609/commits/fc8b356dd347dc035f981a5926f654b1072cbbdb) to direct delegates (just as fast).

That emit approach was different - I passed in `typeof(object)` to the DynamicMethod ctor instead of `objectType` and then `OpCodes.Ldarg_0` (object) + `OpCodes.Ldarg_1` (property value) + `Callvirt` + `Ret`.

I wonder if that was faster than using `Castclass`?

Reply:
You know the answer, right? (: `Castclass` is just a safety check which I added because of discussion in #36506. Remove that operation?

Reply:
I think it's good although a micro benchmark would be useful.

Reply:
If safety isn't so important `Castclass` should be removed.

```ini
BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18895
Intel Xeon CPU E5-2687W 0 3.10GHz, 2 CPU, 8 logical and 8 physical cores
.NET Core SDK=3.0.100-preview5-011568
  [Host]     : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT
  DefaultJob : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT
```

|          Method |     Mean |     Error |    StdDev |          Op/s |
|---------------- |---------:|----------:|----------:|--------------:|
|    WithCastCall | 9.751 ns | 0.3378 ns | 0.9854 ns | 102,552,670.5 |
| WithoutCastCall | 8.346 ns | 0.3414 ns | 1.0068 ns | 119,813,101.1 |

```csharp
public class ValueHolder
{
    public int Value { get; set; }
}

private static readonly Func<object, int> s_WithCastCall = CreatePropertyGetter<ValueHolder, int>(typeof(ValueHolder).GetProperty("Value"), true);
private static readonly Func<object, int> s_WithoutCastCall = CreatePropertyGetter<ValueHolder, int>(typeof(ValueHolder).GetProperty("Value"), false);
private object _holder = new ValueHolder { Value = 42 };

[Benchmark]
public int WithCastCall()
{
    return s_WithCastCall(_holder);
}

[Benchmark]
public int WithoutCastCall()
{
    return s_WithoutCastCall(_holder);
}
```